### PR TITLE
NIAD-108: AMG amendment GP to HA all other

### PR DESCRIPTION
**Pull request by bartek-sarul**

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapper.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    @Override
+    boolean shouldCreateSegment(AmendmentBody amendmentBody) {
+        return StringUtils.isNotBlank(amendmentBody.getFreeText());
+    }
+
+    @Override
+    void validatePatches(JsonPatches patches) {
+        // NOP
```
Comment (Alicja-Kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapper.java:27]: Why there is no validation?

**Review by Alicja-Kainos — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapper.java (hunk 1 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapper.java:27]: Free text is empty or not. Bases on that segment I set segment presence. I think there is nothing more to validate here. However, we'll probably need some `+:?'` escaping but it will have to be done globally

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java [hunk 2]
```diff
@@ -124,10 +127,24 @@ private void validateRequiredFields(AmendmentBody amendmentBody) {
     }
 
     private void validateNonEmptyValues(List<AmendmentPatch> amendmentPatches) {
-        List<String> invalidAmendmentPaths = amendmentPatches.stream()
+        var simpleValuesInvalidPaths = amendmentPatches.stream()
             .filter(amendmentPatch -> amendmentPatch.getValue() != null)
+            .filter(amendmentPatch -> !(amendmentPatch.getValue() instanceof AmendmentExtension))
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java]: Wouldn't `amendmentPatch.isNotExtension()` work here?

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java [hunk 3]
```diff
@@ -124,10 +127,24 @@ private void validateRequiredFields(AmendmentBody amendmentBody) {
...
             .filter(amendmentPatch -> StringUtils.isBlank(amendmentPatch.getFormattedSimpleValue()))
-            .map(AmendmentPatch::getPath)
+            .map(AmendmentPatch::getPath);
+
+        var extensionValuesInvalidPaths = amendmentPatches.stream()
+            .filter(amendmentPatch -> amendmentPatch.getValue() != null)
+            .filter(amendmentPatch -> amendmentPatch.getValue() instanceof AmendmentExtension)
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java]: Wouldn't `amendmentPatch.isExtension()` work here?

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java [hunk 4]
```diff
@@ -124,10 +127,24 @@ private void validateRequiredFields(AmendmentBody amendmentBody) {
...
+            .filter(amendmentPatch -> {
+                var amendmentExtension = (AmendmentExtension) amendmentPatch.getValue();
+                return StringUtils.isNoneBlank(amendmentExtension.getValueBoolean(), amendmentExtension.getValueString());
+            })
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java]:
````java
.filter(amendmentPatch -> StringUtils.isNotBlank(amendmentPatch.getValue().get()))
````

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Override
+    boolean shouldCreateSegment(AmendmentBody amendmentBody) {
+        return amendmentBody.getJsonPatches().getBirthDate().isPresent();
+    }
+
+    @Override
+    void validatePatches(JsonPatches patches) {
+        if (patches.getBirthDate().isPresent()) {
+            if (patches.getBirthDate().get().getOp() == AmendmentPatchOperation.REMOVE) {
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java]: `if (patches.getBirthDate().get().isRemoval()`

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    boolean shouldCreateSegment(AmendmentBody amendmentBody) {
+        return amendmentBody.getJsonPatches().getBirthDate().isPresent();
+    }
+
+    @Override
+    void validatePatches(JsonPatches patches) {
+        if (patches.getBirthDate().isPresent()) {
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java]: Functional way:
````java
if (patches.getBirthDate()
            .map(AmendmentPatch::isRemoval)
            .isPresent()) {	
	throw new PatchValidationException("Illegal remove operation on " + JsonPatches.BIRTH_DATE_PATH));
}
if (patches.getBirthDate()
            .map(AmendmentPatch::getValue)
            .map(AmendmentValue::get)
            .filter(String::isBlank)
            .isPresent()) {
            throw new PatchValidationException("Invalid value for " + JsonPatches.BIRTH_DATE_PATH);
}
````
or use `ifPresent(value -> {throw ...});`

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDrugsDispensedMarkerToEdifactMapper.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Override
+    boolean shouldCreateSegment(AmendmentBody amendmentBody) {
+        return amendmentBody.getJsonPatches().getDrugsDispensedMarker().isPresent();
+    }
+
+    @Override
+    void validatePatches(JsonPatches patches) {
+        patches.getDrugsDispensedMarker()
+            .filter(patch -> patch.getOp() == AmendmentPatchOperation.REMOVE)
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDrugsDispensedMarkerToEdifactMapper.java]: .filter(AmendmentPatch::isRemoval)

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDrugsDispensedMarkerToEdifactMapper.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            .ifPresent(value -> {
+                throw new PatchValidationException("Removing Drugs Dispensed Marker should be done using extension with 'false' value");
+            });
+
+        if (patches.getDrugsDispensedMarker().isPresent()) {
+            if (StringUtils.isBlank(patches.getDrugsDispensedMarker().get().getValue().get())) {
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDrugsDispensedMarkerToEdifactMapper.java]: extract variable

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentResidentialInstituteToEdifactMapper.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    boolean shouldCreateSegment(AmendmentBody amendmentBody) {
+        return amendmentBody.getJsonPatches().getResidentialInstituteCode().isPresent();
+    }
+
+    @Override
+    void validatePatches(JsonPatches patches) {
+        patches.getResidentialInstituteCode()
+            .filter(patch -> patch.getOp() == AmendmentPatchOperation.REMOVE)
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentResidentialInstituteToEdifactMapper.java]: `AmendmentPatch::isRemoval`

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+        var segments = translator.map(amendmentBody);
+
+        assertThat(segments).isPresent().get()
+            .isEqualTo(new FreeText(FREE_TEXT));
+    }
+
+    @Test
+    void whenFreeTextIsEmpty_expectValueIsMapped() {
+        when(amendmentBody.getFreeText()).thenReturn(StringUtils.EMPTY);
```
Comment (grzegorz-robak) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapperTest.java]: Do we need a test for null value?

**Review by grzegorz-robak — CHANGES_REQUESTED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java (hunk 4 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java]: I'll check that

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java (hunk 6 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java]: No, I tried that already. If value is null then it doesn't do the validation as `map` exits in case of null

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapperTest.java (hunk 10 above)
Comment (bartek-sarul) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentFreeTextToEdifactMapperTest.java]: It will be the same but I'll add another test case

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java (hunk 6 above)
Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/translator/amendment/mappers/AmendmentDateOfBirthToEdifactMapper.java]: You're right, then at least extract `patches.getBirthDate().get()` to variable

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java (hunk 4 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/controller/AmendmentController.java]: Can't do that. For `valueBoolean` it can't be null, but for `valueString` it can. I'll leave as is

**Review by grzegorz-robak — DISMISSED** [review]
Consider merging https://github.com/nhsconnect/integration-adaptor-nhais/pull/165

**Review by grzegorz-robak — DISMISSED** [review]

**Review by grzegorz-robak — APPROVED** [review]
